File: libgraffiti/src/selector.rs
// subset of CSS selectors
// x to support CSS-in-JS libs
// x no specificity for now
// x no first/last/nth/siblings
// x universal
// x local name
// x id
// x class
// x child
// x descendant
// x multiple (div, span)
// x combination
// x decoupled from other systems

use crate::util::Atom;

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum Tag {
    LocalName(String),
    Identifier(String),
    ClassNamePart(String),
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Selector {
    parts: Vec<SelectorPart>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum SelectorPart {
    Combinator(Combinator),
    Tag(Atom<Tag>),
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum Combinator {
    Universal,
    Parent,
    Ancestor,
    Or,
}

impl Selector {
    // mask of all tail tags
    // can be stored somewhere and used for short-circuiting:
    // (s.tail_mask().includes(SelectorMask::from(&[Tag::LocalName(..), Tag::ClassNamePart(..), ...])))
    pub fn tail_mask(&self) -> SelectorMask {
        // TODO: split by Combinator::Or and get tail (first because we are reversed)
        //SelectorMask::from(self.parts.iter().filter_map(|p| match p {
        //    SelectorPart::Tag(tag) => Some(tag),
        //    _ => None,
        //}))

        todo!()
    }

    pub fn matches<'a>(&'a self, tags_stack: &[Vec<Atom<Tag>>]) -> bool {
        debug_assert!(tags_stack.len() > 0);

        // useful for reset
        let initial_pos = tags_stack.len() - 1;

        // so we can fast-forward to next OR
        let mut parts_iter = self.parts.iter();

        // state
        let mut pos = initial_pos;
        let mut parent = false;
        let mut ancestors = false;

        // we are always going forward
        'next_part: while let Some(p) = parts_iter.next() {
            match p {
                SelectorPart::Tag(t) => {
                    loop {
                        if parent || ancestors {
                            parent = false;

                            // nothing left to match
                            if pos == 0 {
                                break;
                            }

                            // go up
                            pos -= 1;
                        }

                        if tags_stack[pos].contains(t) {
                            ancestors = false;
                            continue 'next_part;
                        }

                        if !ancestors {
                            break;
                        }
                    }

                    // no match, fast-forward to next OR
                    while let Some(p) = parts_iter.next() {
                        if p == &SelectorPart::Combinator(Combinator::Or) {
                            // reset stack
                            pos = initial_pos;
                            continue 'next_part;
                        }
                    }

                    // or fail otherwise
                    return false;
                }

                // state changes
                SelectorPart::Combinator(Combinator::Parent) => parent = true,
                SelectorPart::Combinator(Combinator::Ancestor) => ancestors = true,

                // no-op
                SelectorPart::Combinator(Combinator::Universal) => {}

                // we still have a match, no need to check others
                SelectorPart::Combinator(Combinator::Or) => return true,
            }
        }

        // everything was fine
        true
    }
}

impl From<&str> for Selector {
    fn from(selector: &str) -> Self {
        parse::selector()
            .parse(selector.trim().as_bytes())
            .expect("invalid selector")
    }
}

pub struct SelectorMask(u32);

impl SelectorMask {
    const BITS: usize = core::mem::size_of::<usize>() * 8;

    pub fn includes(&self, other: SelectorMask) -> bool {
        (self.0 & other.0) != 0
    }
}

impl<'a, T: IntoIterator<Item = &'a Atom<Tag>>> From<T> for SelectorMask {
    fn from(tags: T) -> Self {
        use std::hash::{Hash, Hasher};

        // TODO: maybe it could be oneliner too (fold)
        let hash = |tag: &Tag| {
            let mut hasher = fnv::FnvHasher::default();
            tag.hash(&mut hasher);
            hasher.finish()
        };

        // TODO: test
        // TODO: zero?
        Self(
            tags.into_iter()
                .fold(0, |res, t| res | 1 << (hash(t) as usize - 1) % Self::BITS),
        )
    }
}

mod parse {
    use super::*;
    use pom::char_class::alphanum;
    use pom::parser::*;

    pub fn selector<'a>() -> Parser<'a, u8, Selector> {
        let local_name = ident().map(|s| Tag::LocalName(s.to_string()));
        let id = sym(b'#') * ident().map(|s| Tag::Identifier(s.to_string()));
        let class_name = sym(b'.') * ident().map(|s| Tag::ClassNamePart(s.to_string()));

        // note we parse child/descendant but we flip the final order so it's parent/ancestor
        let universal = sym(b'*').map(|_| Combinator::Universal);
        let child = sym(b' ').repeat(0..) * sym(b'>') * sym(b' ').repeat(0..).map(|_| Combinator::Parent);
        let descendant = sym(b' ').repeat(1..).map(|_| Combinator::Ancestor);
        let or = sym(b' ').repeat(0..) * sym(b',') * sym(b' ').repeat(0..).map(|_| Combinator::Or);

        let tags = (local_name | id | class_name).map(Atom::new).map(SelectorPart::Tag);
        let combinators = (universal | child | descendant | or).map(SelectorPart::Combinator);

        (tags | combinators).repeat(1..).map(|mut parts| {
            // child/descendant -> parent/ancestor
            parts.reverse();

            Selector { parts }
        })
    }

    fn ident<'a>() -> Parser<'a, u8, &'a str> {
        is_a(alphanum_dash).repeat(1..).collect().convert(std::str::from_utf8)
    }

    fn alphanum_dash(b: u8) -> bool {
        alphanum(b) || b == b'-'
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // TODO: it should be possible (tag is nonzero) but maybe it's easier to make Atom shorter
    #[ignore]
    #[test]
    fn part_has_same_size_as_tag() {
        use std::mem::size_of;

        assert_eq!(size_of::<SelectorPart>(), size_of::<Atom<Tag>>())
    }

    #[test]
    fn parsing() {
        use Combinator as C;
        use SelectorPart as P;
        use Tag as T;

        let s = |s| Selector::from(s).parts;
        #[allow(non_snake_case)]
        let A = Atom::new;

        // simple
        assert_eq!(s("*"), &[P::Combinator(C::Universal)]);
        assert_eq!(s("body"), &[P::Tag(A(T::LocalName("body".into())))]);
        assert_eq!(s("h2"), &[P::Tag(A(T::LocalName("h2".into())))]);
        assert_eq!(s("#app"), &[P::Tag(A(T::Identifier("app".into())))]);
        assert_eq!(s(".btn"), &[P::Tag(A(T::ClassNamePart("btn".into())))]);

        // combined
        assert_eq!(
            s(".btn.btn-primary"),
            &[
                P::Tag(A(T::ClassNamePart("btn-primary".into()))),
                P::Tag(A(T::ClassNamePart("btn".into())))
            ]
        );
        assert_eq!(
            s("*.test"),
            &[
                P::Tag(A(T::ClassNamePart("test".into()))),
                P::Combinator(Combinator::Universal)
            ]
        );
        assert_eq!(
            s("div#app.test"),
            &[
                P::Tag(A(T::ClassNamePart("test".into()))),
                P::Tag(A(T::Identifier("app".into()))),
                P::Tag(A(T::LocalName("div".into())))
            ]
        );

        // combined with combinators
        assert_eq!(
            s("body > div.test div#test"),
            &[
                P::Tag(A(T::Identifier("test".into()))),
                P::Tag(A(T::LocalName("div".into()))),
                P::Combinator(C::Ancestor),
                P::Tag(A(T::ClassNamePart("test".into()))),
                P::Tag(A(T::LocalName("div".into()))),
                P::Combinator(C::Parent),
                P::Tag(A(T::LocalName("body".into())))
            ]
        );

        // multi
        assert_eq!(
            s("html, body"),
            &[
                P::Tag(A(T::LocalName("body".into()))),
                P::Combinator(C::Or),
                P::Tag(A(T::LocalName("html".into())))
            ]
        );
        assert_eq!(
            s("body > div, div button span"),
            &[
                P::Tag(A(T::LocalName("span".into()))),
                P::Combinator(C::Ancestor),
                P::Tag(A(T::LocalName("button".into()))),
                P::Combinator(C::Ancestor),
                P::Tag(A(T::LocalName("div".into()))),
                P::Combinator(C::Or),
                P::Tag(A(T::LocalName("div".into()))),
                P::Combinator(C::Parent),
                P::Tag(A(T::LocalName("body".into()))),
            ]
        );
    }

    #[test]
    fn matching() {
        use Tag as T;

        let s = Selector::from;

        let stack = vec![
            vec![T::LocalName("html".into())],
            vec![T::LocalName("body".into()), T::Identifier("app".into())],
            vec![T::LocalName("div".into()), T::Identifier("panel".into())],
            vec![T::LocalName("button".into()), T::ClassNamePart("btn".into())],
            vec![T::LocalName("span".into())],
        ]
        .iter()
        .map(|tags| tags.iter().cloned().map(Atom::new).collect())
        .collect::<Vec<Vec<_>>>();

        // basic
        assert!(s("*").matches(&stack));
        assert!(s("html").matches(&stack[0..1]));
        assert!(s("body").matches(&stack[1..2]));
        assert!(s("div").matches(&stack[2..3]));
        assert!(s("button").matches(&stack[3..4]));
        assert!(s("span").matches(&stack[4..5]));

        // combined
        assert!(s("#app").matches(&stack[1..2]));
        assert!(s("div#panel").matches(&stack[2..3]));
        assert!(s(".btn").matches(&stack[3..4]));

        // parent
        assert!(s("button > span").matches(&stack));
        assert!(s("div#panel > button.btn > span").matches(&stack));

        // ancestor
        assert!(s("button span").matches(&stack));
        assert!(s("div#panel span").matches(&stack));
        assert!(s("body div .btn span").matches(&stack));

        // OR
        assert!(s("div, span").matches(&stack));
        assert!(s("a, b, c, span, d").matches(&stack));
        assert!(s("html, body").matches(&stack[1..2]));

        // complex
        assert!(s("div, span.foo, #panel span").matches(&stack));
        assert!(s("a b c d e f g, span").matches(&stack));
    }
}

File: src/css/selector.zig
// subset of CSS selectors for CSS-in-JS
// TODO: string/symbol interning

const std = @import("std");
const Parser = @import("parser.zig").Parser;
const expectParse = @import("parser.zig").expectParse;
const expectFmt = std.testing.expectFmt;

pub const Specificity = u32;

pub const Selector = struct {
    parts: []const Part,

    const Part = union(enum) {
        // components
        unsupported,
        universal,
        local_name: []const u8,
        identifier: []const u8,
        class_name: []const u8,

        // combinators
        parent,
        ancestor,
        @"or",
    };

    pub fn deinit(self: *Selector, allocator: std.mem.Allocator) void {
        for (self.parts) |part| {
            switch (part) {
                // TODO: intern strings
                .local_name, .identifier, .class_name => |s| allocator.free(s),
                else => {},
            }
        }
        allocator.free(self.parts);
    }

    pub fn format(self: Selector, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
        var i = self.parts.len;
        while (i > 0) {
            i -= 1;

            try switch (self.parts[i]) {
                .unsupported => writer.print(":unsupported", .{}),
                .universal => writer.print("*", .{}),
                .local_name => |s| writer.print("{s}", .{s}),
                .identifier => |s| writer.print("#{s}", .{s}),
                .class_name => |s| writer.print(".{s}", .{s}),
                .parent => writer.print(" > ", .{}),
                .ancestor => writer.print(" ", .{}),
                .@"or" => writer.print(", ", .{}),
            };
        }
    }

    pub fn parse(allocator: std.mem.Allocator, selector: []const u8) !Selector {
        var parser = Parser.init(allocator, selector);

        return try parser.parse(Selector);
    }

    pub fn parseWith(parser: *Parser) !Selector {
        var parts = std.ArrayList(Part).init(parser.allocator);
        errdefer parts.deinit();

        var combinator: ?Part = null;

        while (parser.tokenizer.next() catch null) |tok| {
            if (tok == .lcurly) break;

            const component: ?Part = switch (tok) {
                .star => Part.universal,
                .ident => |s| Part{ .local_name = try parser.allocator.dupe(u8, s) },
                .hash => |s| Part{ .identifier = try parser.allocator.dupe(u8, s) },
                .class_name => |s| Part{ .class_name = try parser.allocator.dupe(u8, s) },
                .colon => Part.unsupported,
                else => null,
            };

            if (component) |comp| {
                if (combinator) |comb| {
                    try parts.append(comb);
                } else if (parser.tokenizer.space_before) {
                    try parts.append(Part.ancestor);
                }

                try parts.append(comp);
                combinator = null;
            } else {
                if (combinator != null) return error.InvalidToken;

                combinator = switch (tok) {
                    .gt => Part.parent,
                    .comma => Part.@"or",
                    .plus => Part.unsupported,
                    .tilde => Part.unsupported,
                    else => return error.InvalidSelectorPart,
                };
            }
        }

        if (parts.items.len == 0 or combinator != null) {
            return error.Eof;
        }

        // save in reverse
        std.mem.reverse(Part, parts.items);

        return .{
            .parts = try parts.toOwnedSlice(),
        };
    }

    pub fn matchElement(self: *const Selector, ctx: anytype, element: anytype) ?Specificity {
        // state
        var i: usize = 0;
        var current = element;
        var parent = false;
        var ancestors = false;
        var specificity: Specificity = 0;

        next_part: while (i < self.parts.len) : (i += 1) {
            switch (self.parts[i]) {
                .parent => parent = true,
                .ancestor => ancestors = true,
                // end-of-branch and we still have a match, no need to check others
                .@"or" => break :next_part,
                else => |comp| {
                    while (true) {
                        if (parent or ancestors) {
                            parent = false;
                            current = ctx.parentElement(current) orelse break;
                        }

                        if (matchComponent(ctx, current, comp)) {
                            ancestors = false;
                            continue :next_part;
                        }

                        // we got no match on parent
                        if (!ancestors) {
                            break;
                        }
                    }

                    // no match, fast-forward to next OR
                    while (i < self.parts.len) : (i += 1) {
                        if (self.parts[i] == .@"or") {
                            // reset stack
                            current = element;
                            continue :next_part;
                        }
                    }

                    // or fail otherwise
                    return null;
                },
            }
        }

        // everything was fine
        return specificity;
    }

    fn matchComponent(ctx: anytype, el: anytype, comp: Part) bool {
        return switch (comp) {
            .universal => true,
            .local_name => |name| std.mem.eql(u8, ctx.localName(el), name),
            .identifier => |id| std.mem.eql(u8, ctx.id(el), id),
            .class_name => |cls| {
                var parts = std.mem.split(u8, ctx.className(el), " ");
                while (parts.next()) |s| if (std.mem.eql(u8, s, cls)) return true;
                return false;
            },
            else => false,
        };
    }
};

test "Selector.format()" {
    try expectFmt("*", "{}", .{Selector{ .parts = &.{.universal} }});
    try expectFmt("div", "{}", .{Selector{ .parts = &.{.{ .local_name = "div" }} }});
    try expectFmt("#app", "{}", .{Selector{ .parts = &.{.{ .identifier = "app" }} }});
    try expectFmt(".btn", "{}", .{Selector{ .parts = &.{.{ .class_name = "btn" }} }});

    try expectFmt("* > *", "{}", .{Selector{ .parts = &.{ .universal, .parent, .universal } }});
    try expectFmt("* *", "{}", .{Selector{ .parts = &.{ .universal, .ancestor, .universal } }});
    try expectFmt("*, *", "{}", .{Selector{ .parts = &.{ .universal, .@"or", .universal } }});
}

fn expectParts(selector: []const u8, parts: []const Selector.Part) !void {
    try expectParse(Selector, selector, Selector{ .parts = parts });
}

test "parsing" {
    // simple
    try expectParts("*", &.{.universal});
    try expectParts("body", &.{.{ .local_name = "body" }});
    try expectParts("h2", &.{.{ .local_name = "h2" }});
    try expectParts("#app", &.{.{ .identifier = "app" }});
    try expectParts(".btn", &.{.{ .class_name = "btn" }});

    // combined
    try expectParts(".btn.btn-primary", &.{ .{ .class_name = "btn-primary" }, .{ .class_name = "btn" } });
    try expectParts("*.test", &.{ .{ .class_name = "test" }, .universal });
    try expectParts("div#app.test", &.{ .{ .class_name = "test" }, .{ .identifier = "app" }, .{ .local_name = "div" } });

    // combined with combinators
    try expectParts("div .btn", &.{ .{ .class_name = "btn" }, .ancestor, .{ .local_name = "div" } });
    try expectParts("body > div.test div#test", &.{
        .{ .identifier = "test" },
        .{ .local_name = "div" },
        .ancestor,
        .{ .class_name = "test" },
        .{ .local_name = "div" },
        .parent,
        .{ .local_name = "body" },
    });

    // multi
    try expectParts("html, body", &.{ .{ .local_name = "body" }, .@"or", .{ .local_name = "html" } });
    try expectParts("body > div, div button span", &.{
        .{ .local_name = "span" },
        .ancestor,
        .{ .local_name = "button" },
        .ancestor,
        .{ .local_name = "div" },
        .@"or",
        .{ .local_name = "div" },
        .parent,
        .{ .local_name = "body" },
    });

    // unsupported for now
    try expectParts(":root", &.{ .{ .local_name = "root" }, .unsupported });
    try expectParts("* + *", &.{ .universal, .unsupported, .universal });
    try expectParts("* ~ *", &.{ .universal, .unsupported, .universal });

    // this should be invalid but it makes rule parsing easier
    try expectParts("* {", &.{.universal});

    // invalid
    try expectParse(Selector, "", error.Eof);
    try expectParse(Selector, " ", error.Eof);
    try expectParse(Selector, "foo + {", error.Eof);
    try expectParse(Selector, "a,,b", error.InvalidToken);
    try expectParse(Selector, "a>>b", error.InvalidToken);

    // TODO: bugs & edge-cases
    // try expectParts("input[type=\"submit\"]", &.{ Unsupported, LocalName("input")]);
}

fn expectMatch(selector: []const u8, index: usize) !void {
    var parser = Parser.init(std.testing.allocator, selector);
    var sel = try parser.parse(Selector);
    defer sel.deinit(std.testing.allocator);

    const parents = [_]?usize{ null, 0, 1, 2, 3 };
    const local_names = [_][]const u8{ "html", "body", "div", "button", "span" };
    const ids = [_][]const u8{ "", "app", "panel", "", "" };
    const class_names = [_][]const u8{ "", "", "", "btn", "" };

    const Ctx = struct {
        fn parentElement(i: usize) ?usize {
            return parents[i];
        }

        fn localName(i: usize) []const u8 {
            return local_names[i];
        }

        fn id(i: usize) []const u8 {
            return ids[i];
        }

        fn className(i: usize) []const u8 {
            return class_names[i];
        }
    };

    try std.testing.expect(sel.matchElement(Ctx, index) != null);
}

test "matching" {
    // basic
    try expectMatch("*", 0);
    try expectMatch("html", 0);
    try expectMatch("body", 1);
    try expectMatch("#app", 1);
    try expectMatch("div", 2);
    try expectMatch("#panel", 2);
    try expectMatch("button", 3);
    try expectMatch(".btn", 3);
    try expectMatch("span", 4);

    // combined
    try expectMatch("body#app", 1);
    try expectMatch("div#panel", 2);
    try expectMatch("button.btn", 3);

    // parent
    try expectMatch("button > span", 4);
    try expectMatch("div#panel > button.btn > span", 4);

    // ancestor
    try expectMatch("button span", 4);
    try expectMatch("div#panel span", 4);
    try expectMatch("html body div button span", 4);
    try expectMatch("body div .btn span", 4);

    // OR
    try expectMatch("div, span", 4);
    try expectMatch("a, b, c, span, d", 4);
    try expectMatch("html, body", 1);

    // complex
    try expectMatch("div, span.foo, #panel span", 4);
    try expectMatch("a b c d e f g, span", 4);
}
